### PR TITLE
[GH-99] Fix JoinURL with encrypted password

### PR DESCRIPTION
**Pull request by Qujja**
#### Summary

Fix the join URL and uses the encrypted password if there is one.

#### Ticket Link

Fixes [#99](https://github.com/mattermost/mattermost-plugin-zoom/issues/99)

### CODE REVIEW

File: server/zoom/meeting.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		MeetingAuthentication        bool   `json:"meeting_authentication"`
+		AuthenticationOption         string `json:"authentication_option"`
+		AuthenticationDomains        string `json:"authentication_domains"`
+		AuthenticationName           string `json:"authentication_name"`
+	} `json:"settings"`
+}
+
+func (c *Client) GetMeeting(meetingID int) (*Meeting, *ClientError) {
+	var ret Meeting
+	return &ret, c.request(http.MethodGet, fmt.Sprintf("/meetings/%v", meetingID), "", &ret)
```
Comment (mickmister) [inline server/zoom/meeting.go:75]: I appreciate the brevity, but can we make this a bit more readable like this?

```suggestion
	var ret Meeting
	err := c.request(http.MethodGet, fmt.Sprintf("/meetings/%v", meetingID), "", &ret)
	return &ret, err
```

File: server/zoom/meeting.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+package zoom
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// The User object defined at https://marketplace.zoom.us/docs/api-reference/zoom-api/meetings/meeting.
+type Meeting struct {
```
Comment (mickmister) [inline server/zoom/meeting.go:12]: Thanks for defining this whole struct!

**Comment by guest7623** [comment]
Hello
Please build fresh release with this fix
I try building on Centos 7, but get error:
`[root@www]# make dist
./build/bin/manifest apply
mkdir -p server/dist;
cd server && env GOOS=linux GOARCH=amd64 /usr/local/go/bin/go build -o dist/plugin-linux-amd64;
cd server && env GOOS=darwin GOARCH=amd64 /usr/local/go/bin/go build -o dist/plugin-darwin-amd64;
cd server && env GOOS=windows GOARCH=amd64 /usr/local/go/bin/go build -o dist/plugin-windows-amd64.exe;
cd webapp && /usr/local/bin/npm run build;

> zoom@1.0.7 build /tmp/mattermost-plugin-zoom/webapp
> webpack --mode=production

Hash: 43bb401670c8b473529d
Version: webpack 4.35.3
Time: 827ms
Built at: 2020-04-20 16:59:37
Entrypoint main =
  [0] external "React" 42 bytes {0} [built]
  [1] external "PropTypes" 42 bytes {0} [built]
 [68] ../locale-data/index.js (ignored) 15 bytes {0} [built]
 [70] external "ReactRedux" 42 bytes {0} [built]
 [71] external "Redux" 42 bytes {0} [built]
 [72] multi ./src/index.js 28 bytes {0} [built]
 [74] (webpack)/buildin/global.js 472 bytes {0} [built]
[103] ./lib/locales (ignored) 15 bytes {0} [built]
[109] ./lib/locales (ignored) 15 bytes {0} [built]
[120] ./src/index.js + 12 modules 73.4 KiB {0} [built]
      | ./src/index.js 1.27 KiB [built]
      | ./src/manifest.js 57 bytes [built]
      | ./src/client/index.js 203 bytes [built]
      | ./src/actions/index.js 1.58 KiB [built]
      | ./src/constants/index.js 154 bytes [built]
      | ./src/client/client.js 1.63 KiB [built]
      | ./src/constants/svgs.js 2.69 KiB [built]
      | ./src/utils/date_utils.js 729 bytes [built]
      |     + 5 hidden modules
    + 111 hidden modules

ERROR in ./src/actions/index.js
Module not found: Error: Can't resolve 'mattermost-redux/action_types' in '/tmp/mattermost-plugin-zoom/webapp/src/actions'
 @ ./src/actions/index.js 4:0-58 45:14-23
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in ./src/client/client.js
Module not found: Error: Can't resolve 'mattermost-redux/client' in '/tmp/mattermost-plugin-zoom/webapp/src/client'
 @ ./src/client/client.js 7:0-50 39:36-43 46:24-31
 @ ./src/client/index.js
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in ./src/client/client.js
Module not found: Error: Can't resolve 'mattermost-redux/client/client4' in '/tmp/mattermost-plugin-zoom/webapp/src/client'
 @ ./src/client/client.js 8:0-62 46:12-23
 @ ./src/client/index.js
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in ./src/components/post_type_zoom/index.js
Module not found: Error: Can't resolve 'mattermost-redux/selectors/entities/common' in '/tmp/mattermost-plugin-zoom/webapp/src/components/post_type_zoom'
 @ ./src/components/post_type_zoom/index.js 10:0-81 18:22-41
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in ./src/index.js
Module not found: Error: Can't resolve 'mattermost-redux/selectors/entities/general' in '/tmp/mattermost-plugin-zoom/webapp/src'
 @ ./src/index.js 6:0-72 28:17-26
 @ multi ./src/index.js

ERROR in ./src/components/post_type_zoom/index.js
Module not found: Error: Can't resolve 'mattermost-redux/selectors/entities/preferences' in '/tmp/mattermost-plugin-zoom/webapp/src/components/post_type_zoom'
 @ ./src/components/post_type_zoom/index.js 9:0-74 17:21-28
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in ./src/components/icon.jsx
Module not found: Error: Can't resolve 'mattermost-redux/utils/theme_utils' in '/tmp/mattermost-plugin-zoom/webapp/src/components'
 @ ./src/components/icon.jsx 5:0-72 23:17-35
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in ./src/components/post_type_zoom/post_type_zoom.jsx
Module not found: Error: Can't resolve 'mattermost-redux/utils/theme_utils' in '/tmp/mattermost-plugin-zoom/webapp/src/components/post_type_zoom'
 @ ./src/components/post_type_zoom/post_type_zoom.jsx 7:0-72 163:17-35
 @ ./src/components/post_type_zoom/index.js
 @ ./src/index.js
 @ multi ./src/index.js

ERROR in chunk main [entry]
main.js
/tmp/mattermost-plugin-zoom/webapp/node_modules/babel-loader/lib/index.js??ref--4!/tmp/mattermost-plugin-zoom/webapp/src/index.js 2f353dc6865657eeb1e6e59137473563
Unexpected token (48:12)
| 
|   const text = await response.text();
|   throw new !(function webpackMissingModule() { var e = new Error("Cannot find module 'mattermost-redux/client/client4'"); e.code = 'MODULE_NOT_FOUND'; throw e; }())(!(function webpackMissingModule() { var e = new Error("Cannot find module 'mattermost-redux/client'"); e.code = 'MODULE_NOT_FOUND'; throw e; }()).url, {
|     message: text || '',
|     status_code: response.status,
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! zoom@1.0.7 build: `webpack --mode=production`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the zoom@1.0.7 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-04-20T13_59_37_923Z-debug.log
make: *** [webapp] Error 2`

**Comment by mickmister** [comment]
@guest7623 Can you try running the following when in the `mattermost-plugin-zoom` directory? It should gather those modules for you. Feel free to report back whether it works or not :+1:

```sh
cd webapp && npm install
```

**Comment by guest7623** [comment]
@mickmister I did it and at the end there were warnings:

> npm WARN zoom@1.0.7 No repository field.
> npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/fsevents):
> npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
> 
> added 747 packages from 370 contributors and audited 10243 packages in 27.413s
> found 13 low severity vulnerabilities
>   run `npm audit fix` to fix them, or `npm audit` for details

As it is now install this plugin in mattermost, simple copy directory mattermost-plugin-zoom to /opt/mattermost/plugins ?

**Comment by mickmister** [comment]
> As it is now install this plugin in mattermost, simple copy directory mattermost-plugin-zoom to /opt/mattermost/plugins ?

@guest7623 This method will require a server restart. Otherwise you can upload via the system console if you have `PluginSettings.EnableUploads` enabled in your config.

**Comment by guest7623** [comment]
@mickmister Sorry, you can in order
1. In /opt/mattermost/plugin i run:
`git clone https://github.com/Qujja/mattermost-plugin-zoom.git`
2. In /opt/mattermost/plugin/mattermost-plugin-zoom/webapp i run:
`npm install`
3. Restart mattermost?
`systemctl restart mattermost`

**Comment by mickmister** [comment]
@guest7623 You're missing the steps of compiling the plugin and installing the plugin. Since you're installing the plugin yourself, I recommend using the upload plugin route. Here's what I suggest:

1. Enable plugin uploads by modifying `/opt/mattermost/config/config.json` to have `EnableUploads: true` in the `PluginSettings` section.
2. Restart the server using `systemctl restart mattermost` so the config setting change is picked up.
3. Clone the repo locally and not on the machine mattermost is running on.
4. `npm install` in `mattermost-plugin-zoom/webapp`
5. `make dist` in `mattermost-plugin-zoom`
6. Go to the system console and upload the file generated at `mattermost-plugin-zoom/dist/zoom/zoom-(version).tar.gz`
7. Enable the plugin if not enabled

No restart should be necessary after uploading the plugin

**Comment by mickmister** [comment]
@guest7623 If you have any other questions, please sign up at https://community.mattermost.com and DM me at `@michael.kochell`

**Comment by mickmister** [comment]
@Qujja Make sure to join the [community server](https://community.mattermost.com) to work more closely with the team! We have a [channel](https://community.mattermost.com/core/channels/plugin-microsoft-calendar) for this project on the server.

**Pull request merged**